_blog/2024-02-07-deploy-rust-web.mdx

```diff
@@ -0,0 +1,89 @@
+
+The basic process involves mostly setting up Nginx or Apache (or a similar proxy), then grabbing the files you need from GitHub or GitLab and compiling the program itself. You can then set the web service up as a systemd service and it will automatically start up whenever the machine starts!
+
+Unfortunately, there are a couple of issues that come along with this: you need to handle all issues yourself (unless the VPS itself has an outage), and typically it’s less scalable than other deployment options. VPS machines typically have a given hard maximum for resource usage and you pay for a set amount - so if you’re not making full usage of the given resources, you may be overpaying. This is especially relevant to Rust, as Rust web applications are typically quite low memory footprint. 
+
+Here’s a list of advantages and disadvantages:
+
+| Advantage | Disadvantage |
+| --- | --- |
+| Total control over all aspects of the machine | If anything goes wrong you need to fix it yourself |
+| You know what you’re getting | Less scalable than serverless - machines typically have a hard limit |
+| Once you learn how to do it, it’s not too hard to do it again | You need to set everything up (Nginx, HTTPS, etc…) |
+| You can run whatever you want (NixOS, Ubuntu, Red Hat Enterprise Linux, etc…) | More expensive than other forms of deployment |
+|  |  |
+
+### Serverless
+
+Serverless nowadays is quite popular and is a great way to host functions that see usage but aren’t run all the time. Typically when developers talk about “serverless”, it is normally referring to serverless functions - code that gets run on servers when a given endpoint is hit and otherwise isn’t run.
+
+Serverless over the years has become quite popular among software developers that want to solve the “scale-to-zero” problem. Services like Cloudflare and AWS both have their own version of “serverless functions” (Cloudflare workers and AWS Lambda, respectively) that simply get run when a HTTP request gets made to a given endpoint. This allows companies to save money by only running the code when required to do so, and is reflected in their pricing - AWS lambda gives you up to 1 million serverless function invocations per month for free!
+
+However, there are some caveats. You often need to adapt your code to the platform so that it can serve your serverless functions. When it comes to Rust, this can typically mean you need to write multiple binaries. You will also be unable to use regular Rust backend frameworks most of the time - though given that one serverless function equates to an endpoint, it’s not too much of a loss. Additionally, cold starts can cause your functions to initially function a little slower.
```

Comment:
Imo, cold starts are the biggest problem.

Reply:
Next is over-engineering - aka having too many serverless functions which should really be one. This hurts performance, cost and maintainability